File: day3/part2.cc
#include <vector>
#include <algorithm>
#include <string>
#include <iostream>
int main() {
   using Value = unsigned long;
   std::vector<Value> values;
   size_t maxbits = 0;
   std::transform(std::istream_iterator<std::string>(std::cin),
                  std::istream_iterator<std::string>(),
                  std::back_inserter(values),
                  [&maxbits](const std::string &s) {
                     maxbits = std::max(maxbits, s.size());
                     return std::stol(s, 0, 2);
                  });
   std::sort(values.begin(), values.end());
   Value out = 1;
   for (int pass = 0; pass < 2; ++pass) {
      auto top = values.begin();
      auto bottom = values.end();
      for (Value key = Value(1) << (maxbits-1); bottom[-1] != *top; key /= 2) {
         auto partition = std::partition_point(top, bottom,
                                    [key] (auto a) { return (a & key) == 0; } );
         if (partition - top <= bottom - partition == bool(pass))
            top = partition;
         else
            bottom = partition;
      }
      out *= *top;
   }
   std::cout << out << std::endl;
}

File: day3/part2.py
import sys
from bisect import bisect_left
with open(sys.argv[1]) if len(sys.argv) > 1 else sys.stdin as f: values = [ int(l.strip(), 2) for l in f ]
values.sort()
maxbits = 1
while 1 << maxbits < values[-1]:
    maxbits+=1
out = 1
for p in range(2):
    top, bottom, shift, key = 0, len(values), maxbits - 1, 0
    while bottom - top > 1:
        key |= 1 << shift
        partition = bisect_left(values, key, top, bottom)
        if (partition - top <= bottom - partition) == (p == 1):
            top = partition
        else:
            bottom = partition
            key &= ~(1 << shift)
        shift -= 1
    out *= values[top]
print(out)
